File: pyfenstein3d/engine/person.py
import math
from .block import Block
from .vector2d import Vector2d
from .field_of_view import FieldOfView
from .config import PERSON_TURN_VELOCITY
from .config import PERSON_MOVEMENT_VELOCITY
from .item_grid import ItemGrid
from .weapon import WeaponPistol

class Person(Block):
    def __init__(self, vector_x: float, vector_y: float, type_id: int, fov: FieldOfView):
        super().__init__(vector_x, vector_y, type_id, True, True)
        self.__fov = fov
        self.__last_pos = self._vector2d.copy()
        self.is_turning_left = False
        self.is_turning_right = False
        self.is_moving_front = False
        self.is_moving_back = False
        self.is_moving_left = False
        self.is_moving_right = False
        self.is_interacting = False
        self._weapon = WeaponPistol()

    @property
    def fov(self):
        return self.__fov

    @property
    def fov_ang(self):
        return self.__fov.ang

    @property
    def weapon(self):
        return self._weapon;

    def update(self, delta_time: float, grid: ItemGrid):
        if self.is_turning_left:
            self.turn(-PERSON_TURN_VELOCITY * delta_time)
        if self.is_turning_right:
            self.turn(PERSON_TURN_VELOCITY * delta_time)
        movement_x = 0
        movement_y = 0
        movement_count = 0
        if self.is_moving_front:
            movement_x += PERSON_MOVEMENT_VELOCITY * delta_time
            movement_count += 1
        if self.is_moving_back:
            movement_x -= PERSON_MOVEMENT_VELOCITY * delta_time
            movement_count += 1
        if self.is_moving_right:
            movement_y += PERSON_MOVEMENT_VELOCITY * delta_time
            movement_count += 1
        if self.is_moving_left:
            movement_y -= PERSON_MOVEMENT_VELOCITY * delta_time
            movement_count += 1
        if movement_x != 0 or movement_y != 0:
            self.move(Vector2d(movement_x / movement_count, movement_y / movement_count))
        if self.is_interacting:
            self.interact(grid)
        self._weapon.update(delta_time, self, grid)

    def can_shoot(self):
        return True

    def shoot(self):
        pass

    def adjust_collision(self, grid: ItemGrid):
        diff_pos = self._vector2d - self.__last_pos
        item = grid.get_block(math.floor(self._vector2d.x + math.copysign(0.5, diff_pos.x)), math.floor(self.__last_pos.y))
        if item is not None and item.is_solid:
            if diff_pos.x > 0:
                self._vector2d = Vector2d(item.block_x - 0.5, self._vector2d.y)
            elif diff_pos.x < 0:
                self._vector2d = Vector2d(item.block_x + 1.5, self._vector2d.y)

        item = grid.get_block(math.floor(self.__last_pos.x), math.floor(self._vector2d.y + math.copysign(0.5, diff_pos.y)))
        if item is not None and item.is_solid:
            if diff_pos.y > 0:
                self._vector2d = Vector2d(self._vector2d.x, item.block_y - 0.5)
            elif diff_pos.y < 0:
                self._vector2d = Vector2d(self._vector2d.x, item.block_y + 1.5)

    def cast(self, grid: ItemGrid):
        self.__fov.cast(self._vector2d, grid)

    def move(self, movement: Vector2d):
        self.__last_pos = self._vector2d.copy()
        self._vector2d = ((self._vector2d ** -self.__fov.ang) + movement) ** self.__fov.ang

    def turn(self, angle: float):
        self.__fov.rot(angle)

    def interact(self, grid: ItemGrid):
        view_dir = (Vector2d.create_with_ang(self.fov.ang) % 0.75) + self._vector2d
        block = grid.get_block(view_dir.x, view_dir.y)
        if block is not None:
            block.interacted()

File: go/engine/person.go
package engine

import "math"

// Person is the base type for all moving characters (player, enemies).
type Person struct {
	block
	fov            *FieldOfView
	lastPos        *Vector2d
	IsTurningLeft  bool
	IsTurningRight bool
	IsMovingFront  bool
	IsMovingBack   bool
	IsMovingLeft   bool
	IsMovingRight  bool
	IsInteracting  bool
	weapon         *Weapon
}

// NewPerson creates a new Person at the given position with the given FieldOfView.
func NewPerson(x, y float64, typeID int, fov *FieldOfView) *Person {
	p := &Person{
		block:  newBlock(x, y, typeID, true, true),
		fov:    fov,
		weapon: NewWeaponPistol().Weapon,
	}
	p.lastPos = p.pos.Copy()
	return p
}

func (p *Person) Fov() *FieldOfView { return p.fov }
func (p *Person) FovAng() float64   { return p.fov.Ang() }
func (p *Person) Weapon() *Weapon   { return p.weapon }

// GetBounds returns bounding points used for FoV culling.
func (p *Person) GetBounds(fovAng float64) []*Vector2d {
	pos := p.pos.Rot(-fovAng)
	return []*Vector2d{
		NewVector2d(pos.X()-0.5, pos.Y()).Rot(fovAng),
		NewVector2d(pos.X()+0.5, pos.Y()).Rot(fovAng),
	}
}

func (p *Person) IsInFov(pos *Vector2d, fovAng, fovDelta, dist float64) bool {
	return blockIsInFov(p.GetBounds(fovAng), pos, fovAng, fovDelta, dist)
}

// Shoot is a no-op for Person (overridden by Player).
func (p *Person) Shoot() {}

// Update advances the person's movement and weapon state.
func (p *Person) Update(deltaTime float64, grid *ItemGrid) {
	if p.IsTurningLeft {
		p.Turn(-PersonTurnVelocity * deltaTime)
	}
	if p.IsTurningRight {
		p.Turn(PersonTurnVelocity * deltaTime)
	}
	movX, movY := 0.0, 0.0
	count := 0
	if p.IsMovingFront {
		movX += PersonMovementVelocity * deltaTime
		count++
	}
	if p.IsMovingBack {
		movX -= PersonMovementVelocity * deltaTime
		count++
	}
	if p.IsMovingRight {
		movY += PersonMovementVelocity * deltaTime
		count++
	}
	if p.IsMovingLeft {
		movY -= PersonMovementVelocity * deltaTime
		count++
	}
	if movX != 0 || movY != 0 {
		p.Move(NewVector2d(movX/float64(count), movY/float64(count)))
	}
	if p.IsInteracting && grid != nil {
		p.Interact(grid)
	}
	p.weapon.Update(deltaTime, p, grid)
}

// AdjustCollision resolves wall collisions after movement.
func (p *Person) AdjustCollision(grid *ItemGrid) {
	diffPos := p.pos.Sub(p.lastPos)

	bx := int(math.Floor(p.pos.X() + math.Copysign(0.5, diffPos.X())))
	by := int(math.Floor(p.lastPos.Y()))
	item := grid.GetBlock(bx, by)
	if item != nil && item.IsSolid() {
		if diffPos.X() > 0 {
			p.setXY(float64(item.BlockX())-0.5, p.pos.Y())
		} else if diffPos.X() < 0 {
			p.setXY(float64(item.BlockX())+1.5, p.pos.Y())
		}
	}

	bx2 := int(math.Floor(p.lastPos.X()))
	by2 := int(math.Floor(p.pos.Y() + math.Copysign(0.5, diffPos.Y())))
	item2 := grid.GetBlock(bx2, by2)
	if item2 != nil && item2.IsSolid() {
		if diffPos.Y() > 0 {
			p.setXY(p.pos.X(), float64(item2.BlockY())-0.5)
		} else if diffPos.Y() < 0 {
			p.setXY(p.pos.X(), float64(item2.BlockY())+1.5)
		}
	}
}

// Cast performs raycasting for this person.
func (p *Person) Cast(grid *ItemGrid) {
	p.fov.Cast(p.asPlayer(), grid)
}

// Move translates the person in the direction of their facing angle.
func (p *Person) Move(movement *Vector2d) {
	p.lastPos = p.pos.Copy()
	p.pos = p.pos.Rot(-p.fov.Ang()).Add(movement).Rot(p.fov.Ang())
	p.block.setXY(p.pos.X(), p.pos.Y())
}

// Turn rotates the person's field of view by the given angle.
func (p *Person) Turn(angle float64) {
	p.fov.Rot(angle)
}

// Interact attempts to interact with the block the person is facing.
func (p *Person) Interact(grid *ItemGrid) {
	viewDir := CreateWithAng(p.fov.Ang()).Norm(0.75).Add(p.pos)
	block := grid.GetBlock(int(viewDir.X()), int(viewDir.Y()))
	if block != nil {
		block.Interacted()
	}
}

// asPlayer returns a *Player view of this Person (needed for Cast).
// Person directly exposes its block data; we cast via a thin wrapper.
func (p *Person) asPlayer() *Player {
	// This is only called when Person is used standalone (tests);
	// actual Player embeds Person so this is not used in production.
	return &Player{Person: p}
}
